blockchain/contracts/NFTtheWorld.sol

```diff
@@ -0,0 +1,180 @@
+
+    mapping(address => uint256[]) public nftReservationInformationOfUsers;
+    mapping(address => uint256[]) public nftAssetsInformationOfUsers;
+
+    // Used to track which addresses have joined the drop
+    mapping(uint256 => address[]) public joinedUsers;
+
+
+    // This function lets a user create a drop by specifiyng a drop time and the number of available NFTs.
+    // During the creation of the drop, the maximum number of NFTs a user can reserve/buy in this drop is calculated.
+    // It is one for a total number of NFTs lower than 20 and 5% otherwise.
+    function createDrop(uint256 _dropTime, uint256 _numberOfNFTS) public {
+        uint256 dropHash = generateRandomNumber(_dropTime);
+        for (uint256 i = 0; i < _numberOfNFTS; i++) {
+            // hardcoded address for testing reasons, to be replaced with msg.sender
+            uint256 nftHash = mockNFT(_dropTime, i + 1, payable(0xC77787e364E3420c0609a249F18De47430900f0C), dropHash);
+            availableNFTs[dropHash].push(nftHash);
+        }
+        // We need to let users buy 1 NFT instead of 5% if there are less than 20
+        if (_numberOfNFTS < 20) {
+            maxNumberOfNFTsToBuy[dropHash] = 1;
+        } else {
+            maxNumberOfNFTsToBuy[dropHash] = (_numberOfNFTS * 5) / 100;
+        }
+        availableNFTsCount[dropHash] = availableNFTs[dropHash].length;
+        dropHashes.push(dropHash);
+        reservedNFTsCount[dropHash] = 0;
+    }
+
+    // Create mock NFTs for drop by creating a list of hashes.
+    function mockNFT(uint256 _dropTime, uint256 _number, address _nftOwner, uint256 _dropHash) internal returns(uint256) {
+        uint256 nftHash = generateRandomNumber(_number);
+        NFTOwnership memory nftOwnership;
+        nftOwnership.nftId = nftHash;
+        nftOwnership.owner = _nftOwner;
+        nftOwnership.reservedFor = _nftOwner;
+        nftOwnership.dropTime = _dropTime;
+        nftOwnership.dropId = _dropHash;
+        nftOwnerships[_dropHash].push(nftOwnership);
+        return nftHash;
+    }
+
+    // This function lets a user join a drop by specifying the number of NFTs she would like to reserve.
+    function joinDrop(uint256 _numberOfNFTs, uint256 _dropHash) public {
+        require(
+            reservedNFTsCount[_dropHash] != availableNFTs[_dropHash].length,
+            "You cannot join the drop anymore."
+        );
+        require(
+            _numberOfNFTs <= maxNumberOfNFTsToBuy[_dropHash],
+            "Sorry, you can't reserve more that 5% of the NFTs."
+        );
+        // We have to make sure that not more NFTs get reserved by users than we have NFTs available
+        require(
+            _numberOfNFTs <= availableNFTs[_dropHash].length - reservedNFTsCount[_dropHash],
+            "Sorry, not enough NFTs left for your request"
+        );
+        reservedNFTsCount[_dropHash] += _numberOfNFTs;
+        nftReservations[msg.sender] = _numberOfNFTs;
```

Comment:
The number of reservations of a user is overwritten with each new join. I.e. if I create a drop, join it 5 times, create a new drop, without dropping the old one, join the new one once, then I have only 1 nftreservation. This likely is unexpected behavior.

Reply:
Good catch, will fix :)